### PR TITLE
send requests that are redirected to devtools

### LINKED ISSUE #11773: Improve network monitor integration
https://github.com/servo/servo/pull/11593 made the network monitor appear when using the [Firefox developer tools](https://github.com/servo/servo/wiki/Devtools) with Servo. There are some known problems with the output that it shows, however:
* we do not report requests that are redirected (like visiting `https://reddit.com` and getting redirected to `https://www.reddit.com`)
* the response panel is empty
* we do not report timing data for "waiting" and "receiving"

I'm not sure how to solve the problem with the response panel yet, but the others will require modifying the code in `http_loader.rs` to send more notifications to the developer tools server. For the missing requests, we should be creating a new request ID each time and sending the information about the request and response sooner. For the missing timing data, we will need to time how long receiving the full HTTP response body takes and send a message to the developer tools that reports it.

Code: `components/net/http_loader.rs`, `components/devtools_traits/lib.rs`, `components/devtools/lib.rs`, `components/devtools/actors/network_event.rs`

cc @mrmiywj This might be interesting to you.

### PR DESCRIPTION
**Pull request by mrmiywj**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11773  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because no automating tests

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11889)

### CODE REVIEW

**Comment by mrmiywj** [comment]
@jdm -reply

**Comment by jdm** [comment]
If you'd like someone in particular to review the changes you can say
r? @jdm
And it will appear in their queue.

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -603,18 +603,17 @@ fn send_request_to_devtools(msg: ChromeToDevtoolsControlMsg,
...
                              status: Option<RawStatus>,
                              pipeline_id: PipelineId) {
-    if let Some(ref chan) = devtools_chan {
-        let response = DevtoolsHttpResponse { headers: headers, status: status, body: None, pipeline_id: pipeline_id };
-        let net_event_response = NetworkEvent::HttpResponse(response);
+    let response = DevtoolsHttpResponse { headers: headers, status: status, body: None, pipeline_id: pipeline_id };
+    let net_event_response = NetworkEvent::HttpResponse(response);
+
+    let msg = ChromeToDevtoolsControlMsg::NetworkEvent(request_id, net_event_response);
+    devtools_chan.send(DevtoolsControlMsg::FromChrome(msg)).unwrap();
```
Comment (izgzhen) [inline components/net/http_loader.rs]: @jdm Is it better to fail silently or `unwrap` here?

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -1066,11 +1107,13 @@ pub fn load<A, B>(load_data: &LoadData,
         // Send an HttpResponse message to devtools with the corresponding request_id
         // TODO: Send this message even when the load fails?
         if let Some(pipeline_id) = load_data.pipeline_id {
+            if devtools_chan.is_some() {
```
Comment (izgzhen) [inline components/net/http_loader.rs]: `let Some` then take reference directly would be tidier? Not sure the motivation of using `as_ref` here ;)

File: components/net/http_loader.rs (hunk 1 above)
Comment (jdm) [inline components/net/http_loader.rs]: Fail silently.

File: components/net/http_loader.rs (hunk 2 above)
Comment (KiChjang) [inline components/net/http_loader.rs]: I agree, using `if let` is more idiomatic here.

**Comment by jdm** [comment]
Sorry for taking so long here! We should write a test for this; there are other devtools+HTTP request tests in tests/unit/net/http_loader.rs (see test_request_and_response_data_with_network_messages). test_load_when_redirecting_from_a_post_should_rewrite_next_request_as_get also shows how to write a test for a redirection.
-S-awaiting-review +S-needs-code-changes +S-needs-tests 

---

Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/http_loader.rs, line 611 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZldE8wS-wX8Lsi6Lc:-KMZldE8wS-wX8Lsi6Ld:-871911173) ([raw file](https://github.com/servo/servo/blob/054bb381e94b82fad4e502bc0f3229f39195c713/components/net/http_loader.rs#L611)):*
> ```Rust
>                              status: Option<RawStatus>,
>                              pipeline_id: PipelineId) {
>     if let Some(ref chan) = devtools_chan {
> ```

I don't see the purpose of this change.

---

*[components/net/http_loader.rs, line 840 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZlJox7xbGQD40uRtc:-KMZl41RADrEPKTuQo7-:-161857662) ([raw file](https://github.com/servo/servo/blob/30851549df5e9dceca2b3dac0ce56ce7c01b50a7/components/net/http_loader.rs#L840)):*
> ```Rust
> struct TFDProvider;
> 
> fn set_request_id(msg: ChromeToDevtoolsControlMsg, id: &String) -> ChromeToDevtoolsControlMsg {
> ```

This function isn't necessary, since we don't need to change the ID.

---

*[components/net/http_loader.rs, line 1047 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZl6XNqHUu-hX4WHL_:-KMZl6XNqHUu-hX4WHLa:2048629579) ([raw file](https://github.com/servo/servo/blob/30851549df5e9dceca2b3dac0ce56ce7c01b50a7/components/net/http_loader.rs#L1047)):*
> ```Rust
>                 //Generate a new request id and send the redirecting request to the
>                 //devtools.
>                 let request_id = uuid::Uuid::new_v4().simple().to_string();
> ```

We don't need to create a new ID here; we already create one each time we make an HTTP request.

---

*[components/net/http_loader.rs, line 1075 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZlRDgjmsluWWDw_pg:-KMZlRDgjmsluWWDw_ph:435558955) ([raw file](https://github.com/servo/servo/blob/30851549df5e9dceca2b3dac0ce56ce7c01b50a7/components/net/http_loader.rs#L1075)):*
> ```Rust
>                         metadata.headers.clone(), metadata.status.clone(),
>                         pipeline_id);
>                 }
> ```

Instead of duplicating all of this code...

---

*[components/net/http_loader.rs, line 1080 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZlnCs6Pa5Z-kbZHWl:-KMZlnCs6Pa5Z-kbZHWm:-321286651) ([raw file](https://github.com/servo/servo/blob/30851549df5e9dceca2b3dac0ce56ce7c01b50a7/components/net/http_loader.rs#L1080)):*
> ```Rust
>         }
> 
>         let mut adjusted_headers = response.headers().clone();
> ```

Let's move all of this code to send the notifications to the devtools to right above the check for whether a redirect will happen. This will avoid the code duplication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11889#-:-KMZktnHhqO0PduuXJ2d:610487175)*

**Comment by jdm** [comment]
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/net/http_loader.rs, line 629 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11889#-KMZmimYa82M9ofD6ADg:-KMZmimZbnTW-S-drSJk:977300527) ([raw file](https://github.com/servo/servo/blob/4089df0ea4d7a8e63d8da9e10e156db587fd60c3/components/net/http_loader.rs#L629)):*
> ```Rust
> 
>     let msg = ChromeToDevtoolsControlMsg::NetworkEvent(request_id, net_event_response);
>     devtools_chan.send(DevtoolsControlMsg::FromChrome(msg));
> ```

Let's use `let _ =` here to avoid the warning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11889)*

**Comment by mrmiywj** [comment]
@jdm 
1. Unit tests have been added
2. I moved the check of devtools_chan from send function into load function, which will make code clean.
3. Other comments have been solved.

**Comment by jdm** [comment]
-S-awaiting-review -S-needs-tests +S-needs-code-changes 

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[tests/unit/net/http_loader.rs, line 640 \[r4\]](https://reviewable.io:443/reviews/servo/servo/11889#-KNgXKkbtn3ZdV9BB_k3:-KNgXKkbtn3ZdV9BB_k4:-352232076) ([raw file](https://github.com/servo/servo/blob/a668c5f952784725a9dacd2683706648239ae3d3/tests/unit/net/http_loader.rs#L640)):*
> ```Rust
>     assert!(devhttprequest.url == url);
>     assert!(devhttpresponse.status == Some(RawStatus(301, Cow::Borrowed("Moved Permanently"))));
> }
> ```

Let's verify that we get two requests and two responses from the devtools listeners - one for the pre-redirect (which we're already checking), and one for the post-redirect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11889#-:-KNgXbljH3NCDHKEcQjs:330330613)*

**Comment by mrmiywj** [comment]
@jdm 
I've added the new tests. 
But, it seems that it fails when compiling with the newest Rust compiler.

**Comment by jdm** [comment]
https://github.com/servo/servo/pull/11889/files#diff-aa469beb5619907dbccd88364264b9b8R1082 needs a `.clone()` to avoid the following errors:
```
/home/travis/build/servo/servo/components/net/http_loader.rs:970:37: 970:49 error: use of moved value: `referrer_url` [E0382]

/home/travis/build/servo/servo/components/net/http_loader.rs:970                                &mut referrer_url);

                                                                                                     ^~~~~~~~~~~~

/home/travis/build/servo/servo/components/net/http_loader.rs:970:37: 970:49 help: run `rustc --explain E0382` to see a detailed explanation

/home/travis/build/servo/servo/components/net/http_loader.rs:970:37: 970:49 note: move occurs because `referrer_url` has type `std::option::Option<url::Url>`, which does not implement the `Copy` trait 

/home/travis/build/servo/servo/components/net/http_loader.rs:1082:29: 1082:41 error: use of moved value: `referrer_url` [E0382]

/home/travis/build/servo/servo/components/net/http_loader.rs:1082         metadata.referrer = referrer_url;

                                                                                              ^~~~~~~~~~~~

/home/travis/build/servo/servo/components/net/http_loader.rs:1082:29: 1082:41 help: run `rustc --explain E0382` to see a detailed explanation

/home/travis/build/servo/servo/components/net/http_loader.rs:1082:29: 1082:41 note: move occurs because `referrer_url` has type `std::option::Option<url::Url>`, which does not implement the `Copy` trait 
```

**Comment by jdm** [comment]
@bors-servo: r+
Yay!

---

Reviewed 2 of 2 files at r5.
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11889#-:-KNhbuEL3L6zq_Qjcu2M:-1396490138)*

**Pull request merged**